### ISSUE TITLE
Store the dest collection as span to a temp for CopyTo optimization

### ISSUE THREAD

**Issue by RikkiGibson** [dotnet/roslyn#71270]
We removed the span temp for now per https://github.com/dotnet/roslyn/pull/71195#discussion_r1425751607.

We'd like to reintroduce the span temp later, perhaps only when non-async, or when lowering can prove that the span will not be used across an await boundary, or perhaps even reinitializing the span temp when it is used across an await boundary.

This would result in a maybe 2% perf improvement for CopyTo codegen.

```cs
T[] result = [..arr1, ..arr2];

// currently becomes
((Span<T>)arr1).CopyTo((Span<T>)result).Slice(...));
((Span<T>)arr2).CopyTo((Span<T>)result).Slice(...));

// we could instead do:
Span<T> spanTemp = result;
((Span<T>)arr1).CopyTo(spanTemp.Slice(...));
((Span<T>)arr2).CopyTo(spanTemp.Slice(...));
```

**Comment by CyrusNajmabadi** [comment]
Yes.  This seems reasonable to do.  Not hugely important.  But why not have better code gen and better perf?
